client/src/fluid/FluidAutocomplete.res

```diff
@@ -436,47 +436,81 @@ let generateExprs = (m: model, props: props, tl: toplevel, ti) => {
-      FPABool(mid, false),
-      FPAConstructor(mid, "Just", list{PBlank(gid())}),
-      FPAConstructor(mid, "Nothing", list{}),
-      FPAConstructor(mid, "Ok", list{PBlank(gid())}),
-      FPAConstructor(mid, "Error", list{PBlank(gid())}),
-      FPANull(mid),
+let generatePatterns = (
+  currentCompletions: list<FT.AutoComplete.data>,
+  ti: tokenInfo,
+  queryString: string,
+): list<item> => {
+  let patternCompletions = List.filterMap(~f=v =>
+    switch v {
+    | {item: FACPattern(p), _} => Some(p)
+    | _ => None
     }
-    |> List.map(~f=p => FT.AutoComplete.FACPattern(p))
-    |> List.filter(~f=c =>
-      // filter out old query string variable
-      switch c {
-      | FT.AutoComplete.FACPattern(FPAVariable(_)) => false
-      | _ => true
+  , currentCompletions)
+
+  let patternOrReplace = (
+    findFn: FT.AutoComplete.patternItem => bool,
+    newPat: FT.AutoComplete.patternItem,
+  ) => List.find(~f=findFn, patternCompletions) |> Option.unwrap(~default=newPat)
+
+  // When possible, re-use an existing pattern rather than generating a new
+  // one. That way, internal IDs are unlikely to change. This is useful as we
+  // re-generate patterns often, including while scrolling in the UI through
+  // the list of available patterns. Each time we regenerate, we attempt to put
+  // highlight the appropriate item - if the ID changes, it becomes difficult
+  // to select the correct item, as simple comparisons won't work.
+  //
+  // Some patterns have no risk of conflict (because they don't have internal
+  // IDs other than the match ID), so we don't bother to prevent conflict.
+  let newStandardPatterns = mid => list{
+    FT.AutoComplete.FPABool(mid, true),
+    FPABool(mid, false),
+    patternOrReplace(p =>
+      switch p {
+      | FPAConstructor(id, "Just", list{PBlank(_)}) => mid == id
+      | _ => false
+      }
+    , FPAConstructor(mid, "Just", list{PBlank(gid())})),
+    FPAConstructor(mid, "Nothing", list{}),
+    patternOrReplace(p =>
+      switch p {
+      | FPAConstructor(id, "Ok", list{PBlank(_)}) => mid == id
+      | _ => false
       }
+    , FPAConstructor(mid, "Ok", list{PBlank(gid())})),
+    patternOrReplace(p =>
+      switch p {
+      | FPAConstructor(id, "Error", list{PBlank(_)}) => mid == id
+      | _ => false
+      }
+    , FPAConstructor(mid, "Error", list{PBlank(gid())})),
+    FPANull(mid),
+  }
+
+  let newVariablePattern = mid => {
+    let matchesExpectedPattern = List.member(
+      ~value=queryString,
+      list{"", "Just", "Nothing", "Ok", "Error", "true", "false", "null"},
     )
 
-  let isInvalidPatternVar = str =>
-    list{"", "Just", "Nothing", "Ok", "Error", "true", "false", "null"} |> List.member(
-      ~value=str,
-    ) || str |> String.dropRight(~count=String.length(str) - 1) |> String.isCapitalized
+    let firstCharacterIsCapitalized =
+      String.dropRight(~count=String.length(queryString) - 1, queryString) |> String.isCapitalized
```

Comment:
Optional: refactor into `String.takeLeft(1)`

Reply:
ha, I knew it'd be something trivial like this. Thanks.